test/dotnet-project-convert.Tests/DotnetProjectConvertTests.cs

```diff
@@ -205,8 +243,353 @@ public void NestedDirectory()
             .EnumerateFileSystemInfos().Select(f => f.Name).Order()
             .Should().BeEquivalentTo(["Program.csproj", "Program.cs"]);
     }
+
+    /// <summary>
+    /// When processing fails due to invalid directives, no conversion should be performed
+    /// (e.g., the target directory should not be created).
+    /// </summary>
+    [Fact]
+    public void ProcessingFails()
+    {
+        var testInstance = _testAssetsManager.CreateTestDirectory();
+        File.WriteAllText(Path.Join(testInstance.Path, "Program.cs"), "#:invalid");
+
+        new DotnetCommand(Log, "project", "convert", "Program.cs")
+            .WithWorkingDirectory(testInstance.Path)
+            .Execute()
+            .Should().Fail()
+            .And.HaveStdErrContaining("Unrecognized directive 'invalid' at");
+
+        new DirectoryInfo(Path.Join(testInstance.Path))
+            .EnumerateDirectories().Should().BeEmpty();
+    }
+
+    /// <summary>
+    /// End-to-end test of directive processing. More cases are covered by faster unit tests below.
+    /// </summary>
+    [Fact]
+    public void ProcessingSucceeds()
+    {
+        var testInstance = _testAssetsManager.CreateTestDirectory();
+        File.WriteAllText(Path.Join(testInstance.Path, "Program.cs"), """
+            #:sdk Aspire.Hosting.Sdk/9.1.0
+            Console.WriteLine();
+            """);
+
+        new DotnetCommand(Log, "project", "convert", "Program.cs")
+            .WithWorkingDirectory(testInstance.Path)
+            .Execute()
+            .Should().Pass();
+
+        new DirectoryInfo(testInstance.Path)
+            .EnumerateFileSystemInfos().Select(f => f.Name).Order()
+            .Should().BeEquivalentTo(["Program"]);
+
+        new DirectoryInfo(Path.Join(testInstance.Path, "Program"))
+            .EnumerateFileSystemInfos().Select(f => f.Name).Order()
+            .Should().BeEquivalentTo(["Program.csproj", "Program.cs"]);
+
+        File.ReadAllText(Path.Join(testInstance.Path, "Program", "Program.cs"))
+            .Should().Be("Console.WriteLine();");
+
+        File.ReadAllText(Path.Join(testInstance.Path, "Program", "Program.csproj"))
+            .Should().Be("""
+                <Project Sdk="Aspire.Hosting.Sdk/9.1.0">
+
+                  <PropertyGroup>
+                    <OutputType>Exe</OutputType>
+                    <TargetFramework>net10.0</TargetFramework>
+                    <ImplicitUsings>enable</ImplicitUsings>
+                    <Nullable>enable</Nullable>
+                  </PropertyGroup>
+
+                </Project>
+
+                """);
+    }
+
+    [Fact]
+    public void Directives()
+    {
+        VerifyConversion(
+            inputCSharp: """
+                #!/program
+                #:sdk Microsoft.NET.Sdk
+                #:sdk Aspire.Hosting.Sdk/9.1.0
+                #:property TargetFramework=net11.0
+                #:package System.CommandLine=2.0.0-beta4.22272.1
```

Comment:
design nit, not to be addressed in this PR: it feels strange that sdk versions and package versions are specified using a different syntax.

Reply:
Assuming you mean the separator (`/` vs `=`), actually either is permitted currently for any directive; each has its own rationale as explained in the spec:

> Any value can optionally have two parts separated by `=` or `/` (the former is consistent with how properties are usually passed, e.g., `/p:Prop=Value`, and the latter is what the `<Project Sdk="Name/Version">` attribute uses).

Reply:
Huh, I have never had a need to specify an sdk version in the project node, so I did not know.

Reply:
It's really only used for NuGet-delivered MSBuild SDKs - common ones are Traversal, NoTargets, and now Cargo, all of which are [here](https://github.com/microsoft/MSBuildSdks).

Reply:
I do resonate with the feedback though - if we have a first-class syntax for the concept, we don't need to adhere to what MSBuild requires necessarily - as long as the 'eject' handles the transformation if required.

Reply:
I can change this easily, I assume `=` would be better?

Reply:
I would personally prefer only allowing the `=`, compared to allowing either `=` or `/`. Otherwise if folks encounter the `/` they may think there is a difference.

If I were starting from scratch I would perhaps even want to just use a space as a delimiter here a la `#pragma`. But, I think that is muddying the waters too much for this PR, I would rather just get something out and get feedback, rather than spending time pursuing that.

Reply:
I think I can do space as the separator, that sounds best to me.

Reply:
> I prefer just changing to `#:package System.CommandLine/2.0.0-beta4.22272.1` and keeping `#:property TargetFramework=net11.0`.

Not sure why would you choose `/` as separator for packages and `=` for properties?

Reply:
> Can you explain why would you choose / as separator for packages and = for properties?

Because that matches the prior-art for each case today. PackageID=PackageVersion is not something seen anywhere else, but PackageID/PackageVersion is, and for properties PropertyName=PropertyValue matches the syntax at the CLI (`/p:Name=Value`).

Reply:
Where is PackageId/PackageVersion used today?